From: Yong-Xuan Wang <yongxuan.wang@sifive.com>
Date: Thu, 14 Mar 2024 14:15:09 +0800
Subject: target/riscv/kvm: fix timebase-frequency when using KVM acceleration
MIME-Version: 1.0
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

The timebase-frequency of guest OS should be the same with host
machine. The timebase-frequency value in DTS should be got from
hypervisor when using KVM acceleration.

Signed-off-by: Yong-Xuan Wang <yongxuan.wang@sifive.com>
Message-ID: <20240314061510.9800-1-yongxuan.wang@sifive.com>
Reviewed-by: Andrew Jones <ajones@ventanamicro.com>
Reviewed-by: Philippe Mathieu-Daudé <philmd@linaro.org>
Reviewed-by: Alistair Francis <alistair.francis@wdc.com>
Signed-off-by: Alistair Francis <alistair.francis@wdc.com>

Origin: backport, https://gitlab.com/qemu-project/qemu/-/commit/385e575cd5ab
X-Backported-By: Heinrich Schuchardt <heinrich.schuchardt@canonical.com>
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/qemu/+bug/2076927
---
 hw/riscv/virt.c              | 2 ++
 target/riscv/kvm/kvm-cpu.c   | 9 +++++++++
 target/riscv/kvm/kvm_riscv.h | 1 +
 3 files changed, 12 insertions(+)

--- qemu.orig/hw/riscv/virt.c	2024-09-25 11:04:14.217243852 -0400
+++ qemu/hw/riscv/virt.c	2024-09-25 11:04:14.214243858 -0400
@@ -743,6 +743,8 @@ static void create_fdt_sockets(RISCVVirt
 
     qemu_fdt_add_subnode(ms->fdt, "/cpus");
     qemu_fdt_setprop_cell(ms->fdt, "/cpus", "timebase-frequency",
+                          kvm_enabled() ?
+                          kvm_riscv_get_timebase_frequency(first_cpu) :
                           RISCV_ACLINT_DEFAULT_TIMEBASE_FREQ);
     qemu_fdt_setprop_cell(ms->fdt, "/cpus", "#size-cells", 0x0);
     qemu_fdt_setprop_cell(ms->fdt, "/cpus", "#address-cells", 0x1);
--- qemu.orig/target/riscv/kvm/kvm-cpu.c	2024-09-25 11:04:14.217243852 -0400
+++ qemu/target/riscv/kvm/kvm-cpu.c	2024-09-25 11:08:59.371876926 -0400
@@ -684,6 +684,16 @@ static void kvm_riscv_put_regs_timer(CPU
     env->kvm_timer_dirty = false;
 }
 
+uint64_t kvm_riscv_get_timebase_frequency(CPUState *cs)
+{
+    uint64_t reg;
+    CPURISCVState *env = &RISCV_CPU(cs)->env;
+
+    KVM_RISCV_GET_TIMER(cs, env, frequency, reg);
+
+    return reg;
+}
+
 typedef struct KVMScratchCPU {
     int kvmfd;
     int vmfd;
--- qemu.orig/target/riscv/kvm/kvm_riscv.h	2024-09-25 11:04:14.217243852 -0400
+++ qemu/target/riscv/kvm/kvm_riscv.h	2024-09-25 11:04:14.214243858 -0400
@@ -27,5 +27,6 @@ void kvm_riscv_aia_create(MachineState *
                           uint64_t guest_num);
 void riscv_kvm_aplic_request(void *opaque, int irq, int level);
 int kvm_riscv_sync_mpstate_to_kvm(RISCVCPU *cpu, int state);
+uint64_t kvm_riscv_get_timebase_frequency(CPUState *cs);
 
 #endif
